simphony/cuds/model.py

```diff
@@ -39,119 +55,179 @@ def __init__(self):
         # that will return the corresponding value.
         self._map = {}
 
-    @staticmethod
-    def _is_cuds_component(obj):
-        """Check whether the object is a CUDS component."""
-        # When I see a bird that walks like a duck and swims like a duck
-        # and quacks like a duck, I call that bird a duck
-        # TODO: replace with metadata type checks
-        return all([hasattr(obj, 'name'),
-                    hasattr(obj, 'data')])
+        # Another map to keep a mapping between names and ids
+        self._name_to_id_map = {}
+
+    @property
+    def uid(self):
+        return self._uid
+
+    @uid.setter
+    def uid(self, value):
+        raise TypeError("uid is readonly.")
+
+    @uid.deleter
+    def uid(self):
+        raise TypeError("Can't delete uid.")
 
     @staticmethod
     def _is_dataset(obj):
         """Check if the object is a dataset."""
-        return isinstance(obj, (ABCParticles, ABCLattice, ABCMesh))
+        return isinstance(obj, (ABCParticles,
+                                ABCLattice,
+                                ABCMesh,
+                                api.DataSet))
 
     @property
     def data(self):
         """Data container for CUDS items."""
         return DataContainer(self._data)
```

Comment:
You are correct, this will not work properly with our current mutable CUDS objects. I think the behavior of the `data` property regarding meta classes is unclear.

I think `Simulation` can have the same API as `CUDS`.